src/lib/appValidation.ts

```diff
@@ -0,0 +1,435 @@
+      break;
+    case AjvErrorKeyword.Required:
+      onlyKeyNode = true;
+      message = errorMessages.validation.missingRequiredField(
+        error.params.missingProperty
+      );
+      break;
+    case AjvErrorKeyword.Type:
+      message = `Incorrect type: ${error.message}`;
+      break;
+    default:
+      break;
+  }
+  return { message, nodePath, onlyKeyNode };
+}
+
+function diagnosticDeduplicationKey(
+  nodePath: (string | number)[],
+  keyword: string,
+  params: any
+): string {
+  return `${keyword}::${nodePath.join('/')}::${JSON.stringify(params)}`;
+}
+
+function getKeyNode(node: jsonc.Node): jsonc.Node | undefined {
+  return node.parent?.type === 'property' ? node.parent.children?.[0] : node;
+}
+
+async function getCachedSchemasForAllTypes(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number
+): Promise<SchemaCacheEntry> {
+  // If schema cache has a schema for this version, return it
+  // If there is an error, we're going to return that too
+  if (schemaCache[platformVersion]) {
+    return schemaCache[platformVersion];
+  }
+  let response: Record<string, AnySchema>;
+  try {
+    response = await getIntermediateRepresentationSchema({
+      projectSourceDir: projectRoot.fsPath,
+      platformVersion,
+      accountId,
+    });
+    let byType: Record<string, SchemaAndValidator> = {};
+    for (const [key, schema] of Object.entries(response)) {
+      byType[key] = { schema };
+    }
+    const result = {
+      byType,
+    };
+    schemaCache[platformVersion] = result;
+    return result;
+  } catch (error) {
+    const message = `${errorMessages.api.failedToFetchSchemas}. Reload your window to try again`;
+    console.error(message, error);
+    const result = {
+      error: message,
+    };
+    schemaCache[platformVersion] = result;
+    return result;
+  }
+}
+
+async function getCachedValidatorForExternalType(
+  projectRoot: Uri,
+  platformVersion: string,
+  accountId: number,
+  externalType: string
+): Promise<ValidateFunction | { error: string }> {
+  const cachedSchemas = await getCachedSchemasForAllTypes(
+    projectRoot,
+    platformVersion,
+    accountId
+  );
+  if ('error' in cachedSchemas) {
+    return cachedSchemas;
+  }
+  const internalType = mapToInternalType(externalType);
+  if (!(internalType in cachedSchemas.byType)) {
+    return { error: errorMessages.validation.unsupportedType(externalType) };
+  }
+  const schemaAndValidator = cachedSchemas.byType[internalType];
+  let validator = schemaAndValidator.validator;
+  if (!validator) {
+    validator = ajv.compile(schemaAndValidator.schema);
+    schemaAndValidator.validator = validator;
+  }
+  return validator;
+}
+
+// VSCode doesn't provide a dirname function, that's in the github:/microsoft/vscode-uri package
+function vscodeDirname(uri: Uri): Uri {
+  return uri.with({ path: dirname(uri.path) });
+}
+
+async function findProjectRoot(start: Uri): Promise<Uri | null> {
```

Comment:
Could we use `vscode.workspace.findFiles` for this? I'm not very familiar with how it works or exactly what it does, but I see it being used [here](https://github.com/HubSpot/hubspot-cms-vscode/blob/2a4498c36039e8b5efde26dfa20c993dc6fa61cd/src/lib/providers/fileCompletion.ts#L7)

Reply:
Yeah looks like there's no built-in way to do that. We could potentially use it to locate every `hsproject.json` file, and then filter that list down to just the paths where `currentDir.path.startsWith(potentialParentHsProjectJSON.path)` is true. But apparently findFiles will look through the entire workspace, which isn't as performant as this. I'm surprised they don't have a better built in way to do it.

What you have here seems like the best way to do it. Maybe we could make it a little easier to follow by turning the try/catch for `fs.stat` into a dedicated "doesFileExist()` util? There's an existing fileHelpers file that it could live in.